psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -118,62 +119,84 @@ void DirectJK::print_header() const {
         outfile->Printf("    Screening Type:    %11s\n", screen_type.c_str());
         outfile->Printf("    Screening Cutoff:  %11.0E\n", cutoff_);
         outfile->Printf("    Incremental Fock:  %11s\n", incfock_ ? "Yes" : "No");
+        if (incfock_) outfile->Printf("    Fixed reference:   %11s\n", fixed_reference_ ? "Yes" : "No");
         outfile->Printf("\n");
     }
 }
 void DirectJK::preiterations() {
-
 #ifdef USING_BrianQC
     if (brianEnable) {
-        double threshold = cutoff_ * (brianCPHFFlag ? 1e-3 : 1e-0); // CPHF needs higher precision
+        double threshold = cutoff_ * (brianCPHFFlag ? 1e-3 : 1e-0);  // CPHF needs higher precision
         brianCOMSetPrecisionThresholds(&brianCookie, &threshold);
         checkBrian();
     }
 #endif
 }
 
 void DirectJK::incfock_setup() {
-    if (do_incfock_iter_) {
-        size_t njk = D_ao_.size();
+    // Sanity check
+    size_t njk = D_ao_.size();
+    if (D_reference_.size() != njk) do_incfock_iter_ = false;
 
-        // If there is no previous pseudo-density, this iteration is normal
-        if (initial_iteration_ || D_prev_.size() != njk) {
-	        initial_iteration_ = true;
-
-            D_ref_ = D_ao_;
-            zero();
-        } else { // Otherwise, the iteration is incremental
-            for (size_t jki = 0; jki < njk; jki++) {
-                D_ref_[jki] = D_ao_[jki]->clone();
-                D_ref_[jki]->subtract(D_prev_[jki]);
-            }
+    if (do_incfock_iter_) {
+        // Incremental formation
+        for (size_t jki = 0; jki < njk; jki++) {
+            D_current_[jki] = D_ao_[jki]->clone();
+            D_current_[jki]->subtract(D_reference_[jki]);
         }
     } else {
-        D_ref_ = D_ao_;
+        // Full formation
+        D_current_ = D_ao_;
         zero();
     }
 }
 
 void DirectJK::incfock_postiter() {
-    // Save a copy of the density for the next iteration
-    D_prev_.clear();
-    for(auto const &Di : D_ao_) {
-        D_prev_.push_back(Di->clone());
+    // Add in the reference J and K
+    if (do_incfock_iter_) {
+        size_t njk = D_ao_.size();
+        for (size_t jki = 0; jki < njk; jki++) {
+            if (do_J_) J_ao_[jki]->add(J_reference_[jki]);
+            if (do_K_) K_ao_[jki]->add(K_reference_[jki]);
+            if (do_wK_) wK_ao_[jki]->add(wK_reference_[jki]);
+        }
+
+        // If we don't do incremental formation, we can update the reference
```

Comment:
The main thing I'd like to point out is right here. With this implementation, the update of the reference is tied to the `INCFOCK_FULL_FOCK_EVERY` keyword, because the reference will only be updated when an IncFock iteration isn't done, i.e., when the full Fock matrix is constructed. This is, of course, great for numerical stability. But, to get consistent reference updates, it forces you to set `INCFOCK_FULL_FOCK_EVERY` to a smallish value. Psi4 uses a small default value already (5), but one goal of the IncFock changes is to avoid consistent reconstruction of the Fock matrix (for some comparisons, ORCA has a similar keyword set to 15 by default if I recall, and GAMESS never recomputes the full Fock matrix). 

Would it be good to decouple the regeneration of the reference from the formation of the full Fock matrices? With such an implementation, some references could be difference matrices themselves, which could be a  numerical stability concern. But perhaps this is okay in practice?

Reply:
I'm not sure I follow what you say, but I'll have another look at the code with clearer eyes when I have the time.

One thing one could do with the changes is to use the fixed-reference variant in the background even when one uses traditional incremental formation: the "full" rebuild would be done with respect to the fixed reference matrix, which would make the consecutive full rebuilds much faster, as one would only need to compute the full matrix once.